native-dash-app-example/src/app.py

```diff
@@ -0,0 +1,217 @@
+#
+# (C) Copyright 2010-2022 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This file and its contents are confidential information and NOT open source.
+# Distribution is prohibited.
+
+import datetime
+import logging
+import os
+import sys
+from functools import partial, wraps
+from urllib.parse import unquote
+
+import pandas as pd
+import plotly.express as px
+import requests
+from dash import dash, dcc, html
+from dash.dependencies import Input, Output
+from flask import Flask, make_response, redirect, request, session
+from jupyterhub.services.auth import HubOAuth
+from jupyterhub.utils import isoformat
+
+from flask_session import Session
+
+# When running flask in debug mode outside of a JupyterHub environment,
+# deactivate the `authenticated` and `track_activity` decorator.
+FLASK_DEBUG = int(os.environ.get("FLASK_DEBUG", 0))
+
+LOG = logging.getLogger(__name__)
+formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
+handler = logging.StreamHandler(sys.stdout)
+handler.setFormatter(formatter)
+handler.setLevel(logging.DEBUG)
+LOG.addHandler(handler)
+LOG.setLevel(logging.INFO)
+if FLASK_DEBUG:
+    LOG.setLevel(logging.DEBUG)
+
+# When run from Edge, these environment variables will be provided
+PREFIX = os.environ.get("JUPYTERHUB_SERVICE_PREFIX", "/")
+API_TOKEN = os.environ.get("JUPYTERHUB_API_TOKEN", "")
+ACTIVITY_URL = os.environ.get("JUPYTERHUB_ACTIVITY_URL", None)
+SERVER_NAME = os.environ.get("JUPYTERHUB_SERVER_NAME", "")
+API_URL = os.environ.get("JUPYTERHUB_API_URL", "http://127.0.0.1:8081")
+APP_VERSION = os.environ.get("APP_VERSION", "native-dash-app-example")
+EDGE_API_SERVICE_URL = os.environ.get("EDGE_API_SERVICE_URL", None)
+EDGE_API_ORG = os.environ.get("EDGE_API_ORG", None)
+
+DASH_URL_BASE_PATHNAME = "dashboard/"
+
+AUTH = HubOAuth(api_token=API_TOKEN, cache_max_age=60, api_url=API_URL)
+
+LOG.debug(f"JUPYTERHUB_SERVER_NAME {SERVER_NAME}")
+LOG.debug(f"JUPYTERHUB_SERVICE_PREFIX {PREFIX}")
+LOG.debug(f"JUPYTERHUB_ACTIVITY_URL {ACTIVITY_URL}")
+
+
+def track_activity(f):
+    """Decorator for reporting server activities with the Hub"""
+
+    @wraps(f)
+    def decorated(*args, **kwargs):
+        if FLASK_DEBUG == 1:
+            return f(*args, **kwargs)
+        last_activity = isoformat(datetime.datetime.now())
+        if ACTIVITY_URL:
+            try:
+                requests.post(
+                    ACTIVITY_URL,
+                    headers={
+                        "Authorization": f"token {API_TOKEN}",
+                        "Content-Type": "application/json",
+                    },
+                    json={"servers": {SERVER_NAME: {"last_activity": last_activity}}},
+                )
+            except Exception:
+                pass
+        return f(*args, **kwargs)
+
+    return decorated
+
+
+def view_authenticated(view_func, **view_args):
+    """Wraps the view function with authentication."""
+    token = session.get("token")
```

Comment:
It may be better to use dcc.Store rather than Flask session, since that is the Dash native way of storing state.

Reply:
I guess it depends if we are targeting Dash-knowledgable users or Flask-knowledgable users.

Reply:
Apparently if I use the layout func, I have to use dcc.Store since it's outside the Flask requests context.

Reply:
I think I got the state working (maybe, perhaps) but in the authentication it fails because we are not in a flask request context while trying to generate the oauth2 redirect:

`state = AUTH.generate_state(next_url=request.path)`

=>

```
  File "/home/app_admin/native-dash-app-example/app.py", line 147, in create_app
    app.layout = authenticated_layout
  File "/home/app_admin/.edm/envs/edm/lib/python3.8/site-packages/dash/dash.py", line 636, in layout
    layout_value = self._layout_value()
  File "/home/app_admin/.edm/envs/edm/lib/python3.8/site-packages/dash/dash.py", line 596, in _layout_value
    layout = self._layout() if self._layout_is_function else self._layout
  File "/home/app_admin/native-dash-app-example/app.py", line 128, in authenticated_layout
    state = AUTH.generate_state(next_url=request.path)
  File "/home/app_admin/.edm/envs/edm/lib/python3.8/site-packages/werkzeug/local.py", line 316, in __get__
    obj = instance._get_current_object()  # type: ignore[misc]
  File "/home/app_admin/.edm/envs/edm/lib/python3.8/site-packages/werkzeug/local.py", line 513, in _get_current_object
    raise RuntimeError(unbound_message) from None
RuntimeError: Working outside of request context.
```